src/sentry/ingest/transaction_clusterer/tree.py

```diff
@@ -0,0 +1,157 @@
+
+/users
+  /*
+    /posts
+      *
+    /settings
+
+Resulting in the following replacement rules:
+
+/users/*/posts/*  # To replace both identifiers
+/users/*          # For URLs with only one identifier, e.g. /users/*/settings
+
+"""
+
+import logging
+from collections import UserDict, defaultdict
+from typing import Iterable, List, Optional, Union
+
+import sentry_sdk
+from typing_extensions import TypeAlias
+
+from .base import Clusterer, ReplacementRule
+
+__all__ = ["TreeClusterer"]
+
+
+class Merged:
+    pass
+
+
+#: Symbol representing a merged node
+MERGED = Merged()
+
+#: Separator by which we build the tree
+SEP = "/"
+
+
+logger = logging.getLogger(__name__)
+
+
+class TreeClusterer(Clusterer):
+    def __init__(self, *, merge_threshold: int) -> None:
+        self._merge_threshold = merge_threshold
+        self._graph = Node()
+
+    def add_input(self, transaction_names: Iterable[str]) -> None:
+        for tx_name in transaction_names:
+            parts = tx_name.split(SEP)
+            node = self._graph
+            for part in parts:
+                node = node.setdefault(part, Node())
+
+    def get_rules(self) -> List[ReplacementRule]:
+        """Merge high-cardinality nodes in the graph and extract rules"""
+        with sentry_sdk.start_span(op="txcluster_merge"):
+            self._graph.merge(self._merge_threshold)
+
+        if self._graph.count_leaf_nodes() == 1:
+            logger.warning("Graph collapsed into a single path, merge threshold might be too low")
+            return []
+
+        # Generate exactly 1 rule for every merge
+        rule_paths = [path for path in self._graph.paths() if path[-1] is MERGED]
+
+        # TODO: Warn if top level collapse
+        # (in production, we should not allow rules that merge the top-level URL segments)
+
+        # Sort by path length, descending (most specific rule first)
+        rule_paths.sort(key=len, reverse=True)
+
+        return [self._build_rule(path) for path in rule_paths]
+
+    @staticmethod
+    def _build_rule(path: List["Edge"]) -> ReplacementRule:
+        path_str = SEP.join(["*" if isinstance(key, Merged) else key for key in path])
+        path_str += "/**"
+        return ReplacementRule(path_str)
+
+
+#: Represents the edges between graph nodes. These edges serve as keys in the
+#: node dictionary.
+Edge: TypeAlias = Union[str, Merged]
+
+
+class Node(UserDict):  # type: ignore
+    """Keys in this dict are names of the children"""
+
+    def paths(self, ancestors: Optional[List[Edge]] = None) -> Iterable[List[Edge]]:
+        """Collect all paths and subpaths through the graph"""
+        if ancestors is None:
+            ancestors = []
+        for name, child in self.items():
+            path = ancestors + [name]
+            yield path
+            yield from child.paths(ancestors=path)
+
+    @property
+    def is_leaf(self) -> bool:
+        return not self.values()
```

Comment:
I was confused by this at first. When building a trie for cases like this, a "leaf" node and a "terminal" node is not the same. A "leaf" node is always a node that has no children, whereas a "terminal" node is a node where a path ends. Consider the following example:

```txt
/user/bob/settings
/user/bob/settings/security
```

`settings` is a terminal node but not a leaf node, and `security` is both a leaf and terminal node. However, this is not relevant at all for the algorithm, since the rule generated by the longer path is a "super rule" of all the rules of the shorter paths.

I'm saying this because I think this is worth writing somewhere. It may ease any debugging we need to do in some time.

Reply:
Removed this property now, it was only used by the "single path" safety check.